huaweicloud/services/cae/resource_huaweicloud_cae_component.go

```diff
@@ -0,0 +1,524 @@
+	}
+
+	return []map[string]interface{}{
+		{
+			"replica":        utils.PathSearch("replica", spec, nil),
+			"runtime":        utils.PathSearch("runtime", spec, nil),
+			"source":         flattenSource(utils.PathSearch("source", spec, nil)),
+			"resource_limit": flattenResourceLimit(utils.PathSearch("resource_limit", spec, nil)),
+			"build":          flattenBuild(parameters, utils.PathSearch("build", spec, nil)),
+		},
+	}
+}
+
+func flattenResourceLimit(resourceLimit interface{}) []map[string]interface{} {
+	return []map[string]interface{}{
+		{
+			"cpu":    utils.PathSearch("cpu_limit", resourceLimit, nil),
+			"memory": utils.PathSearch("memory_limit", resourceLimit, nil),
+		},
+	}
+}
+
+func flattenSource(source interface{}) []map[string]interface{} {
+	if source == nil {
+		return nil
+	}
+
+	return []map[string]interface{}{
+		{
+			"type":     utils.PathSearch("type", source, nil),
+			"url":      utils.PathSearch("url", source, nil),
+			"sub_type": utils.PathSearch("sub_type", source, nil),
+			"code":     flattenCode(utils.PathSearch("code", source, nil)),
+		},
+	}
+}
+
+func flattenCode(code interface{}) []map[string]interface{} {
+	if (code) == nil {
+		return nil
+	}
+	return []map[string]interface{}{
+		{
+			"auth_name": utils.PathSearch("auth_name", code, nil),
+			"branch":    utils.PathSearch("branch", code, nil),
+			"namespace": utils.PathSearch("namespace", code, nil),
+		},
+	}
+}
+
+func flattenBuild(parameters interface{}, build interface{}) []map[string]interface{} {
+	archive := utils.PathSearch("archive", build, nil)
+	if archive == nil {
+		return nil
+	}
+
+	return []map[string]interface{}{
+		{
+			"archive": []map[string]interface{}{
+				{
+					"artifact_namespace": utils.PathSearch("artifact_namespace", archive, nil),
+				},
+			},
+			"parameters": parameters,
+		},
+	}
+}
+
+func resourceComponentUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg     = meta.(*config.Config)
+		region  = cfg.GetRegion(d)
+		httpUrl = "v1/{project_id}/cae/applications/{application_id}/components/{component_id}"
+		product = "cae"
+	)
+	client, err := cfg.NewServiceClient(product, region)
+	if err != nil {
+		return diag.Errorf("error creating CAE Client: %s", err)
+	}
+
+	updatePath := client.Endpoint + httpUrl
+	updatePath = strings.ReplaceAll(updatePath, "{project_id}", client.ProjectID)
+	updatePath = strings.ReplaceAll(updatePath, "{application_id}", d.Get("application_id").(string))
+	updatePath = strings.ReplaceAll(updatePath, "{component_id}", d.Id())
+
+	updateComponentOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		MoreHeaders: map[string]string{
+			"X-Environment-Id": d.Get("environment_id").(string),
+		},
+		JSONBody: utils.RemoveNil(buildCreateOrUpdateComponentBodyParams(d)),
+	}
+	_, err = client.Request("PUT", updatePath, &updateComponentOpt)
+	if err != nil {
+		return diag.Errorf("error updating CAE component (%s): %s", d.Id(), err)
```

Comment:
```suggestion
	updatePath = strings.ReplaceAll(updatePath, "{component_id}", componentId)

	updateComponentOpt := golangsdk.RequestOpts{
		KeepResponseBody: true,
		MoreHeaders: map[string]string{
			"X-Environment-Id": d.Get("environment_id").(string),
		},
		JSONBody: utils.RemoveNil(buildCreateOrUpdateComponentBodyParams(d)),
	}
	_, err = client.Request("PUT", updatePath, &updateComponentOpt)
	if err != nil {
		return diag.Errorf("error updating CAE component (%s): %s", componentId, err)
```
suggest to define a variable and names 'componentId'